[Jaryl Goh Jun Zhong] iP

rionshocker opened the pull request:
# Rion

> Wisdom is knowing what to do next; virtue is doing it. - David Starr Jordan

The Rion chatbot serves as a bot that helps you to remember what you need to do. It's

- text-based
- easy to learn
- _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/rionshocker/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😆

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's my `main` method:
```
package duke;

import duke.command.Command;

/**
 * Shows the chat bot, Rion.
 */
public class Duke {
    /** Storage object that deals with loading and saving tasks. */
    private Storage storage;
    /** The unique task list containing all tasks inputted. */
    private TaskList taskList;
    /** The Ui that deals with interactions with the users. */
    private Ui ui;

    /**
     * Creates the chatbot with a specific file path.
     * It checks whether there is a cached version of the list.
     * The resulting task list will then be saved at the path.
     * 
     * @param filePath Path for the task list to be stored in.
     */
    public Duke(String filePath) {
        this.storage = new Storage(filePath);
        this.ui = new Ui();

        try {
            taskList = new TaskList(storage.loadList());
        } catch (DukeException e) {
            ui.showLoadingError(e.getMessage());
            this.taskList = new TaskList();
        }
    }

    /**
     * Runs the chatbot, Rion.
     * It receives commands and executes them accordingly.
     */
    private void runRion() {
        boolean isExit = false;
        while (!isExit) {
            try {
                String fullCommand = ui.nextCommand();
                Command c = Parser.parse(fullCommand);
                c.execute(taskList, ui, storage);
                isExit = c.isExit();
            } catch (DukeException e) {
                ui.printError(e.getMessage());
            }
        }
    }

    
    public static void main(String[] args) {
        Duke rion = new Duke("./data/tasks.txt");
        rion.runRion();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

rionshocker committed:
Level-0

rionshocker committed:
Added echo function

rionshocker committed:
Add the ability to add and list tasks entered.

rionshocker committed:
Added a class to represent tasks and its functions.

rionshocker committed:
Add support for tracking of todos,deadlines,events

rionshocker committed:
Added Automated Text UI Testing.

rionshocker committed:
Added handling errors.

rionshocker committed:
Add support for deleting tasks.

rionshocker committed:
Updated expected output.

rionshocker committed:
Add the ability to save and load list

rionshocker committed:
Add the ability to understand dates and times

rionshocker committed:
Make the code more OOP

rionshocker committed:
Organize into Packages

rionshocker committed:
Merge branch 'add-gradle-support' into A-Gradle

rionshocker committed:
Add JUnit Tests

rionshocker committed:
Package the App as a JAR file

rionshocker committed:
Add JavaDoc comments

rionshocker committed:
Add the find function of tasks

rionshocker committed:
Add the find function for tasks

jamz903 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,311 @@
+            String input = command.replaceFirst("mark", "").trim();
+            int index = validateMarkIndex(input);
+            return new MarkCommand(index);
+        } else if (isUnMarkCommand(command)) {
+            String input = command.replaceFirst("unmark", "").trim();
+            int index = validateMarkIndex(input);
+            return new UnMarkCommand(index);
+        } else if (isAddToDoCommand(command)) {
+            String todo = command.replaceFirst("todo", "").trim();
+            validateToDo(todo);
+            return new AddToDoCommand(todo);
+        } else if (isAddDeadlineCommand(command)) {
+            String[] deadline = validateDeadline(command);
+            if (isDate(deadline[1])) {
+                LocalDateTime deadlineDate = parseDate(deadline[1]);
+                return new AddDeadlineCommand(deadline[0], deadlineDate);
+            } else {
+                return new AddDeadlineCommand(deadline[0], deadline[1]);
+            }
+        } else if (isAddEventCommand(command)) {
+            String[] event = validateEvent(command);
+            if (isDate(event[1])) {
+                LocalDateTime eventDate = parseDate(event[1]);
+                LocalDateTime endTime = parseDate(event[2]);
+                return new AddEventCommand(event[0], eventDate, endTime);
+            } else {
+                return new AddEventCommand(event[0], event[1], event[2]);
+            }
+        } else if (isDeleteCommand(command)) {
+            String input = command.replaceFirst("delete", "").trim();
+            int index = validateDeleteIndex(input);
+            return new DeleteCommand(index);
+        } else if (isFindCommand(command)) {
+            String keyword = command.replace("find", "").trim();
+            validateFindInput(keyword);
+            return new FindCommand(keyword);
+        } else {
+            throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
```
This method is really long! You could try to refactor your code by breaking it down into multiple sub-methods for better readability.

jamz903 reviewed (commented):
Good job! Your implementation is extremely comprehensive and well done 👍

rionshocker committed:
Add a GUI to the chatbot

peiran18 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,311 @@
+     * @return an array of String objects that represents a Deadline.
+     * @throws DukeException If input is invalid or incomplete.
+     */
+    private static String[] validateDeadline(String input) throws DukeException {
+        if (input.equals("deadline")) {
+            throw new DukeException("OOPS!! You need to include the deadline!");
+        }
+        String description = input.replace("deadline ", "");
+        String[] details = description.split(" /by ");
+        if (details.length != 2) {
+            throw new DukeException("OOPS!! You need to include the deadline!");
+        }
+        return details;
+    }
```
I really like how you go the extra mile to check for instances like when the user forgets to input the date!

peiran18 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+        taskList.get(index).changeStatus();
+    }
+
+    /**
+     * Adds the task to the task list.
+     * 
+     * @param task The task to be added to the list.
+     */
+    public void add(Task task) {
+        taskList.add(task);
+    }
```
Following your markTask(), maybe name this method ass addTask() to provide more information and also be more consistent? The same follows for delete etc

peiran18 commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,12 @@
+deadline return book /by Sunday
+event project meeting /from Mon 2pm /to 4pm
+deadline do homework /by no idea :-p
+bye
```
I like how you have so many commands! Don't forget to check umark/find as well?

peiran18 reviewed (commented):
Overall I think your code is very clear and addresses a lot of edge case! The javadocs look great too :)

rionshocker committed:
Add assertions to test committed

There are several methods that rely on assumptions that certain method
calls will work. As such, the assertions aim to test the assumptions.

rionshocker committed:
Improve code quality

Following the 'Code Quality' topics, code is edited to make it more
readable.

rionshocker committed:
Merge pull request #1 from rionshocker/branch-A-Assertions

Add assertions to test

rionshocker committed:
Merge pull request #2 from rionshocker/branch-A-CodeQuality

Improve code quality

rionshocker committed:
Merge branch 'branch-A-CodeQuality'

rionshocker committed:
Merge commit 'd43b443c55bf61cad871e49bb2cb4e5164f2b13c'

rionshocker committed:
Add the ability to update tasks easily

rionshocker committed:
Update features

rionshocker committed:
Remove text file

rionshocker committed:
Add a representative screenshot

rionshocker committed:
Add a user guide

rionshocker committed:
Update code quality

rionshocker committed:
Update UG and UI

rionshocker committed:
Update commands

rionshocker committed:
Update the update command

rionshocker committed:
Update UG

rionshocker committed:
Add additional test
